### PR TITLE
resolve bad migration operation causing drop of permissions (fixes #403)

### LINKED ISSUE #403: [BUG] Many existing service permissions on group anonymous lost after upgrading from 1.7.3 to 3.5.1
**Describe the bug**
Test upgrade of production instance for PR https://github.com/bird-house/birdhouse-deploy/pull/107. This is done by importing the Magpie Postgres DB from prod `pavics.ouranos.ca` into staging `medus.ouranos.ca`.  

In this case, the upgrade failed on imported DB from prod `pavics.ouranos.ca`.

However, `medus.ouranos.ca` itself was successfully upgraded via the autodeploy and no errors found.

My dev VM was also upgraded and no Jenkins errors, except the bug found in https://github.com/Ouranosinc/Magpie/issues/401 where the Anonymous group, Thredds config page completely crash.

Basically, it seems like the upgrade is a random hit-or-miss depending on the existing Magpie data.  3 different environments upgraded with 3 completely different results.

Thredds perm before

crop-Screenshot_2021-03-25 Magpie Administration-boreas-be4-gr-anon-thredds

Thredds perm after (lost the perm on the top level "thredds")

crop-Screenshot_2021-03-25 Magpie-boreas-after-upgrade-group-anon-thredds

WPS perm before

crop-Screenshot_2021-03-25 Magpie Administration-boreas-be4-anon-group-wps

WPS perm after

crop-Screenshot_2021-03-25 Magpie-boreas-after-upgrade-group-anon-wps png

All the other services used to have all permissions turned on before and all lost after upgrade

crop-Screenshot_2021-03-25 Magpie-boreas-after-upgrade-group-anon-geoserverwms png

crop-Screenshot_2021-03-25 Magpie-boreas-after-upgrade-group-anon-ncwms png

crop-Screenshot_2021-03-25 Magpie-boreas-after-upgrade-group-anon-wfs png

Jenkins test suite effectively failed corresponding to those permissions lost.

Thredds perms lost:
```
OSError: [Errno -78] NetCDF: Authorization failure: b'https://medus.ouranos.ca/twitcher/ows/proxy/thredds/dodsC/birdhouse/cccma/CanESM2/historical/day/atmos/r1i1p1/tasmin/tasmin_day_CanESM2_historical_r1i1p1_18500101-20051231.nc'
```

WPS perms lost:
```
16:08:21  ServiceException: <?xml version="1.0" encoding="utf-8"?>
16:08:21  <ExceptionReport version="1.0.0"
16:08:21      xmlns="http://www.opengis.net/ows/1.1"
16:08:21      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
16:08:21      xsi:schemaLocation="http://www.opengis.net/ows/1.1 http://schemas.opengis.net/ows/1.1.0/owsExceptionReport.xsd">
16:08:21      <Exception exceptionCode="NoApplicableCode" locator="AccessForbidden">
16:08:21          <ExceptionText>Not authorized to access this resource. User does not meet required permissions.</ExceptionText>
16:08:21      </Exception>
16:08:21  </ExceptionReport>
```

**To Reproduce**
1. Save `/data/magpie_persist` folder from prod `pavics.ouranos.ca`: `cd /data; tar czf magpie_persist.prod.tgz magpie_persist`
2. scp `magpie_persist.prod.tgz` to `medus`
3. login to `medus`
4. `cd /path/to/birdhouse-deploy/birdhouse`
3. `./pavics-compose.sh down`
4. `git checkout 1.11.24`  # before PR https://github.com/bird-house/birdhouse-deploy/pull/107
5. `cd /data`
2. `rm -rf magpie_persist`
3. `tar xzf magpie_persist.prod.tgz`  # restore Magpie DB with prod version
4. `cd /path/to/birdhouse-deploy/birdhouse`
5. `./pavics-compose.sh up -d`
3. Update `env.local` `MAGPIE_ADMIN_PASSWORD` with prod passwd for Twitcher to be able to access Magpie since unable to change admin passwd back to the old value on `medus`, issue https://github.com/Ouranosinc/Magpie/issues/402
4. `./pavics-compose.sh restart twitcher`  # for Twitcher to get new Magpie admin passwd 
4. Baseline working state: trigger Jenkins test suite, ensure all pass except `pavics_thredds.ipynb` that requires new Magpie
5. Baseline working state: take screen shot of existing services permissions on group Anonymous
6. `git checkout` [7be3eeaf0464511abc89e93d1f553de2d7413b59](https://github.com/bird-house/birdhouse-deploy/commit/7be3eeaf0464511abc89e93d1f553de2d7413b59)  # Pr merge commit of new Magpie
7. `./pavics-compose.sh up -d`  # upgrade to new Magpie
8. `docker logs magpie`: check no DB migration error
9. Trigger Jenkins test suite again
10. Take screen shot of new services permissions on group Anonymous

**Expected behavior**
Should not lose existing permissions, Jenkins test suite should have same result as before upgrade.

### PR DESCRIPTION
**Pull request by fmigneault**
resolve bad migration operation causing drop of permissions (fixes #403)

### CODE REVIEW

File: magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py [hunk 1]
```diff
@@ -43,7 +43,7 @@ def upgrade():
     grp_res_perms = GroupResourcePermissionService.base_query(db_session=session)
     usr_res_perms = UserResourcePermissionService.base_query(db_session=session)
 
-    for perm_list in [grp_res_perms, usr_res_perms]:
+    for item_id_key, perm_list in [("group_id", grp_res_perms), ("user_id", usr_res_perms)]:
```
Comment (tlvu) [inline magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py:46]: I have no idea about DB migration but shouldn't the fix be in a new migration step?  Fixing in an existing migration step means a server that already execute this step won't attempt it again and miss out on the fix?

Discard what I said if it makes no sense since I know nothing about DB migration.

**Review by tlvu — APPROVED** [review]
Approve the spirit of it as I have no knowledge of DB migration.

Curious, how do you know this will fix the issue since I do not see any new test case reproducing the issue to confirm the fix works?

I assume once this PR is merged, you'll create a new Magpie docker and test a migration in a real PAVICS stack using our birdhouse-deploy recipe?

File: magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py (hunk 1 above)
Comment (fmigneault) [reply magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py:46]: It has to be applied there so that a server that did not run the migration will do it properly after rollback.
If added as new step, permissions will still be dropped at that step and later 'new' step will have no effect as there is nothing left to work with.
Migration steps have their own version dissociated from magpie versions.

**Comment by fmigneault** [comment]
> Curious, how do you know this will fix the issue since I do not see any new test case reproducing the issue to confirm the fix works?
> 
> I assume once this PR is merged, you'll create a new Magpie docker and test a migration in a real PAVICS stack using our birdhouse-deploy recipe?

@tlvu 
I used birdhouse-deploy version 1.11.24 and applied all permissions seen in #401, then ran the migration operations via debug script to step through lines. I noticed that the permission resolution was incorrectly matching bad permissions together based on that the missing compare this PR adds. 
After patch, the permissions seem to remain as expected. (still need to test on your side though, as I don't have the full database)

File: magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py (hunk 1 above)
Comment (tlvu) [reply magpie/alembic/versions/2020-09-22_a2a039e2cff5_explicit_permission_names.py:46]: > It has to be applied there so that a server that did not run the migration will do it properly after rollback.

Oh wait so there is a way to rollback a DB migration?  I didn't have to `rm -rf /data/magpie_persist` when reverting from 3.5.1 to 1.7.3?

**Comment by tlvu** [comment]
> I used birdhouse-deploy version 1.11.24 and applied all permissions seen in #401,

You meant #403?  Not #401 right?

**Comment by fmigneault** [comment]
> You meant #403? Not #401 right?
@tlvu 

Yes. My bad. 
I tagged a new version 3.8.0 with all fixes. 
You can let me know if you see any more errors.

**Comment by tlvu** [comment]
> I tagged a new version 3.8.0 with all fixes.
> You can let me know if you see any more errors.

@fmigneault 

Woo a jump from 3.5.1 to 3.8.0?  So during last Magpie upgrade, 3.7.0 was available but we didn't want to use it?

Before I spend, and potentially waste time on this 3.8.0 again, what tests did you perform on this 3.8.0?  Is it integrated here https://github.com/bird-house/birdhouse-deploy/compare/restore-previous-broken-magpie-upgrade-so-we-can-work-on-a-fix, how has it been tested?

**Comment by dbyrns** [comment]
@tlvu, Francis test the migration to 3.8.0 using this procedure : https://github.com/Ouranosinc/Magpie/pull/407#issuecomment-808524273.
So this migration is tested and safe. However, no migration should be done to an earlier version as, like we discussed with @huard, we don't want to support old versions and the fix is in 3.8.0. The only thing we cannot do is testing with your own server data, so hopefully you will not waste your time this time. Also remember that Magpie has an exhaustive list of tests, but edge cases with production data are hard to predict and it's not because of ill will that errors happen.
